aten/src/ATen/native/sparse/cuda/SparseCUDABlas.cu

```diff
@@ -71,6 +80,119 @@ cusparseOperation_t convertTransToCusparseOperation(char trans) {
+
+  if (csrvala == nullptr || b == nullptr || c == nullptr) return; 
+
+  cusparseOperation_t opa = convertTransToCusparseOperation(transa);
+  cusparseOperation_t opb = convertTransToCusparseOperation(transb);
+
+  // cusparseSpMM actually supports int64_t. 
+  // In order to support int64 here, index pointers csrrowptra, csrcolinda have to be passed as int64_t. 
+  TORCH_CHECK((m <= INT_MAX) && (n <= INT_MAX) && (k <= INT_MAX) && (nnz <= INT_MAX) && (ldb <= INT_MAX) && (ldc <= INT_MAX),
+    "By now, cusparseSpMM only supports m, n, k, nnz, ldb, ldc with the bound [val] <= ", INT_MAX, ".", 
+    "Please submit an issue on Github."
+  );
+
+  int64_t ma = m, ka = k; 
+  if (transa != 'n') std::swap(ma, ka); 
+
+  cusparseSpMatDescr_t descA; 
+  TORCH_CUDASPARSE_CHECK(cusparseCreateCsr(
+    &descA,                     /* output */
+    ma, ka, nnz,                /* rows, cols, number of non zero elements */
+    csrrowptra,                 /* row offsets of the sparse matrix, size = rows +1 */
+    csrcolinda,                 /* column indices of the sparse matrix, size = nnz */
+    csrvala,                    /* values of the sparse matrix, size = nnz */
+    CUSPARSE_INDEX_32I,         /* data type of row offsets index */
+    CUSPARSE_INDEX_32I,         /* data type of col indices */
+    CUSPARSE_INDEX_BASE_ZERO,   /* base index of row offset and col indes */
+    cusparse_value_type         /* data type of values */
+  )); 
+
+  int64_t kb = k, nb = n;
+  if (transb != 'n') std::swap(kb, nb); 
+
+  cusparseDnMatDescr_t descB; 
+  TORCH_CUDASPARSE_CHECK(cusparseCreateDnMat(
+    &descB,               /* output */
+    kb, nb, ldb,          /* rows, cols, leading dimension */
+    b,                    /* values */
+    cusparse_value_type,  /* data type of values */
+    CUSPARSE_ORDER_COL    /* memory layout, ONLY column-major is supported now */
+  )); 
+
+  cusparseDnMatDescr_t descC; 
+  TORCH_CUDASPARSE_CHECK(cusparseCreateDnMat(
+    &descC,               /* output */
+    m, n, ldc,            /* rows, cols, leading dimension */
+    c,                    /* values */ 
+    cusparse_value_type,  /* data type of values */ 
+    CUSPARSE_ORDER_COL    /* memory layout, ONLY column-major is supported now */
+  )); 
+
+
+  auto handle = at::cuda::getCurrentCUDASparseHandle();
+
+  // cusparseSpMM_bufferSize returns the bufferSize that can be used by cusparseSpMM
+  size_t bufferSize; 
+  TORCH_CUDASPARSE_CHECK(cusparseSpMM_bufferSize(
+    handle, opa, opb,     
+    &alpha,               
+    descA, descB, 
+    &beta, 
+    descC, 
+    cusparse_value_type,  /* data type in which the computation is executed */
+    CUSPARSE_CSRMM_ALG1,  /* default computing algorithm for CSR sparse matrix format */
+    &bufferSize           /* output */
+  )); 
+
+  void* externalBuffer; // device pointer
+  AT_CUDA_CHECK(cudaMalloc(&externalBuffer, bufferSize)); 
```

Comment:
It'll be better for performance if you route this buffer through the CUDA caching allocator so we don't have to sync here. Something like:

```
      auto& allocator = *c10::cuda::CUDACachingAllocator::get();
      buffer_ = allocator.allocate(size);
```

The ensuing DataPtr will also take care of deallocation automatically when you're done too

Reply:
fixed